designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,784 @@
+
+```
+your-project
+├── eslint.config.js
+└── node_modules
+  ├── eslint
+  ├── eslint-config-a
+  | └── node_modules
+  |   └── eslint-plugin-example@1.0.0
+  └── eslint-config-b
+    └── node_modules
+      └── eslint-plugin-example@2.0.0
+```
+
+The problem comes when the shareable configs try to use the default namespace of `eslint-plugin-example` for its rules, such as:
+
+```js
+module.exports = {
+    ruledefs: {
+        ...require("eslint-plugin-example").ruledefs
+    }
+};
+```
+
+If both shareable configs do this, and the user tries to use both shareable configs, an error will be thrown because the `ruledefs` namespace "example" can only be assigned once.
+
+To work around this problem, shareable configs that rely on plugins should also export a separate config that assigns a unique namespace to all plugin rules. To continue with the example in this section, that might be under `eslint-config-example/compat` and would look like this:
+
+```js
+// pull in ruldefs from the regular config
+const originalRuleDefs = require("./index").ruledefs;
+
+// namespace each original ruledef
+const compatRuleDefs = {};
+
+for (const key in originalRuleDefs) {
+    compatRuleDefs["config-a::" + key] = originalRuleDefs[key];
+}
+
+// include in a config
+module.exports = {
+    ruledefs: {
+        ...compatRuleDefs
+    }
+};
+```
+
+Here, the shareable config is exporting the same rules from the plugins as before, but using the namespace `config-a::example` instead of the default `example` that `eslint-plugin-example` defines. In this way, shareable config authors can provide an option for users who may be extending multiple configs that depend on the same plugin.
+
+If a shareable config does not provide a compat version, then the end user still can create one on their own by manually creating a new config from the shareable config. For example:
+
+```js
+// get the config you want to extend
+const configToExtend = require("eslint-plugin-example");
+
+// create a new copy
+const compatConfig = Object.create(configToExtend);
+compatConfig.ruledefs = {};
+
+// namespace each original
+for (const key in configToExtend.ruledefs) {
+    compatConfig.ruledefs["compat::" + key] = configToExtend.ruledefs[key];
+}
+
+// include in config
+exports.config = [
+    compatConfig,
+    {
+        // overrides here
+    }
+];
+```
+
```

Comment:
Could we document which of the following changes would constitute breaking changes for a shareable config, taking into account the possibility of causing errors from ruledef conflicts?

1. Adding a new plugin (ruledef) to a shareable config, without enabling any of its rules
1. Configuring a rule from an existing ruledef which previously was not configured
1. Adding more rules to an existing ruledef (e.g. by upgrading a plugin), without configuring the rules

My understanding is that (1) would be a breaking change because it could cause a new ruledef conflict for the config's users. (2) would be a breaking change for other reasons (it would usually cause more errors to be reported for end users). (3) depends on whether we decide to allow "conflicting" ruledefs that have the same object identity.

Reply:
Can we simplify this by saying that any change to a shareable config's `ruledefs` is a breaking change?

Reply:
Possibly, although making (3) a breaking change seems undesirable (e.g. it would include bugfixes in plugins). Unless a shareable config pinned its plugin dependencies to an exact version, a "breaking change" like this could occur accidentally if a new version of a plugin is released. This is a disadvantage of having behavior changes dependent on how package managers flatten packages.

Reply:
I don't want to get too far into prescribing when shareable configs should consider a new release a breaking change (we don't do this today and arguably already have the same problem). I'd rather just warn shareable config authors that *any* change, whether that be to a dependency or to what they're outputting is possibly a breaking change and leave it at that. (Arguably absolutely any change will affect people's builds.)